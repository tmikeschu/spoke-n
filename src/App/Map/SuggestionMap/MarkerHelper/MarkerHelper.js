// @flow
import React from "react"
import { Marker } from "react-google-maps"
import { categoryIcons } from "../../category_data"

interface LatLngable {
  lat: number,
  lng: number
}

interface Locationable {
  location: LatLngable,
  id: number,
  category: string
}

interface Positionable {
  position: LatLngable
}

export const coordinatesFor = (point: Locationable): Positionable => {
  return {
    position: {
      lat: parseFloat(point.location.lat),
      lng: parseFloat(point.location.lng),
    }
  }
}

export const currentLocationPin = (currentLocation: Locationable): Marker => (
  <Marker
    {...coordinatesFor(currentLocation)}
    animation={2}
    icon={categoryIcons["cycling"]}
    options={{clickable: true}}
  />
)


export const suggestionMarkers = (suggestions: Locationable[], handler: Function): Marker[] => (
  suggestions.map((suggestion, i) => {
    const marker: Positionable = coordinatesFor(suggestion)

    return (
      <Marker
        key={suggestion.id}
        icon={categoryIcons[suggestion.category]}
        {...marker}
        onClick={(props) => handler(props)}
      />
    )
  })
)

export const lineCoordinates = (points: Locationable[]): LatLngable[] => (
  points.map(point => (
    {
      lat: parseFloat(point.location.lat),
      lng: parseFloat(point.location.lng)
    }
  ))
)

export const endsOfDayMarkers = (actualPath: Locationable[]): Marker[] => (
  actualPath.slice(0, actualPath.length - 1)
    .map((point, i) => {
      const marker = coordinatesFor(point)

      return (
        <Marker
          key={point.id}
          icon={categoryIcons["endOfDay"]}
          {...marker}
        >
        </Marker>
      )
    })
)

export const suggestion = (suggestionPin: LatLngable): Marker => (
  suggestionPin.lat !== undefined && (
    <Marker
      options={{ clickable: true}}
      position={suggestionPin}/>
  )
)

