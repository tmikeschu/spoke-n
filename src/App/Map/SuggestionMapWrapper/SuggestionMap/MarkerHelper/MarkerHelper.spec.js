import React, { Component } from 'react'
import { Marker } from "react-google-maps"
import * as MarkerHelper from './MarkerHelper'

describe('MarkerHelper', () => {
  const point = {
    location: {
      lat: "40.1",
      lng: "30.9",
    },
    category: "stay"
  }

  const point2 = {
    location: {
      lat: "41.0",
      lng: "20.9",
    }
  }

  const expectedPosition = {
    position: {
      lat: 40.1,
      lng: 30.9
    }
  }

  describe('#coordinatesFor', () => {
    it('returns a position object with lat lng', () => {
      const coordinates = MarkerHelper.coordinatesFor(point)
      expect(coordinates).toMatchObject(expectedPosition)
    })
  })

  describe('#currentLocationPin', () => {
    it('returns a JSX marker', () => {
      const currentLocationPin = MarkerHelper.currentLocationPin(point)
      expect(currentLocationPin.type.displayName).toEqual("Marker")

      expect(currentLocationPin.props).toBeTruthy
      const { position, icon, options, animation } = currentLocationPin.props
      expect(position).toEqual(expectedPosition["position"])
      expect(icon).toEqual("https://maps.google.com/mapfiles/ms/icons/cycling.png")
      expect(animation).toEqual(2)
      expect(options).toMatchObject({clickable: true})
    })
  })

  describe('#suggestionMarkers', () => {
    it('returns a collection of markers', () => {
      const suggestions = [point, point2]
      const handler = () => {}
      const markers = MarkerHelper.suggestionMarkers(suggestions, handler)
      const allMarkers = markers.every(m => m.type.displayName === "Marker")
      expect(allMarkers).toBeTruthy()

      const marker = markers[0]

      const { position, icon, onClick } = marker.props
      expect(position).toEqual(expectedPosition["position"])
      expect(icon).toEqual("https://maps.google.com/mapfiles/ms/icons/blue.png")
      expect(onClick.name).toEqual("onClick")
    })
  })

  describe('#lineCoordinates', () => {
    it('returns a collection of latLng objects', () => {
      const points = [point, point2]
      const line = MarkerHelper.lineCoordinates(points)
      const allLatLng = line.every(c => c.lat && c.lng)
      expect(allLatLng).toBeTruthy()
      expect(line[0].lat).toEqual(40.1)
      expect(line[0].lng).toEqual(30.9)
    })
  })

  describe('#endsOfDayMarkers', () => {
    it('returns a collection of markers', () => {
      const points = [point, point2]
      const markers = MarkerHelper.endsOfDayMarkers(points)
      const allMarkers = markers.every(m => m.type.displayName === "Marker")
      expect(allMarkers).toBeTruthy()

      const marker = markers[0]
      const { position, icon } = marker.props
      expect(position).toEqual(expectedPosition["position"])
      expect(icon).toEqual("https://maps.google.com/mapfiles/ms/icons/flag.png")
    })
  })

  describe('#suggestion', () => {
    it('returns a marker', () => {
      const suggestion = point.location 
      const marker = MarkerHelper.suggestion(suggestion)
      expect(marker.type.displayName).toEqual("Marker")

      const { position, options } = marker.props
      const expected = { lat: "40.1", lng: "30.9" }
      expect(position).toMatchObject(expected)
      expect(options).toMatchObject({ clickable:  true })
    })
  })
})

