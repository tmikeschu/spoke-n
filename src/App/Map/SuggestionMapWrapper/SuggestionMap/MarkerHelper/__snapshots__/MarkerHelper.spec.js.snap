// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkerHelper #currentLocationPin returns a JSX marker 1`] = `
<Marker
  animation={2}
  icon="none"
  label={
    Object {
      "fontSize": "48px",
      "text": "🚲",
    }
  }
  options={
    Object {
      "clickable": true,
    }
  }
  position={
    Object {
      "lat": 40.1,
      "lng": 30.9,
    }
  }
/>
`;

exports[`MarkerHelper #endsOfDayMarkers returns a collection of markers 1`] = `
<Marker
  fill="transparent"
  icon="none"
  label={
    Object {
      "fontSize": "24px",
      "text": "🏳",
    }
  }
  position={
    Object {
      "lat": 40.1,
      "lng": 30.9,
    }
  }
/>
`;

exports[`MarkerHelper #suggestion returns a marker 1`] = `
<Marker
  icon="none"
  label={
    Object {
      "fontSize": "32px",
      "text": "📍",
    }
  }
  options={
    Object {
      "clickable": true,
    }
  }
  position={
    Object {
      "lat": "40.1",
      "lng": "30.9",
    }
  }
/>
`;

exports[`MarkerHelper #suggestionMarkers returns a collection of markers 1`] = `
<Marker
  icon="none"
  label={
    Object {
      "fontSize": "32px",
      "text": "🏡",
    }
  }
  onClick={[Function]}
  position={
    Object {
      "lat": 40.1,
      "lng": 30.9,
    }
  }
/>
`;

exports[`MarkerHelper #suggestionMarkers returns a collection of markers 2`] = `
<Marker
  icon="none"
  label={
    Object {
      "fontSize": "32px",
      "text": undefined,
    }
  }
  onClick={[Function]}
  position={
    Object {
      "lat": 41,
      "lng": 20.9,
    }
  }
/>
`;
